programs/cipher/pbcrypt.c

```diff
@@ -0,0 +1,720 @@
+    memset( header, 0, 4 );
+    header += 4;
+#endif
+    *header++ = ( metadata->payload_size >> 24 ) & 0xff;
+    *header++ = ( metadata->payload_size >> 16 ) & 0xff;
+    *header++ = ( metadata->payload_size >> 8 ) & 0xff;
+    *header++ = ( metadata->payload_size ) & 0xff;
+
+    return( ret );
+}
+
+#define DESERIALIZE_CHECK( size, expr, cond, message )  \
+    do                                                  \
+    {                                                   \
+        (void) ( expr );                                \
+        if( ! ( cond ) )                                \
+        {                                               \
+            mbedtls_fprintf( stderr, "%s\n", message ); \
+            return( PBCRYPT_ERR );                      \
+        }                                               \
+        header += ( size );                             \
+    }                                                   \
+    while( 0 )
+#define DESERIALIZE( size, expr )               \
+    DESERIALIZE_CHECK( size, expr, 1, "" )
+static int pbcrypt_deserialize_header( pbcrypt_metadata_t *metadata,
+                                       unsigned char *header )
+{
+    unsigned long iterations;
+    unsigned long long payload_size;
+    DESERIALIZE_CHECK( sizeof( magic ), 0,
+                       memcmp( header, magic, sizeof( magic ) ) == 0,
+                       "File format error: invalid magic header" );
+    DESERIALIZE_CHECK( 1, 0, *header == 0,
+                       "File format error: invalid version" );
+    DESERIALIZE( 1, metadata->md = mbedtls_md_info_from_type( *header ) );
+    DESERIALIZE( 1,
+                 metadata->cipher = mbedtls_cipher_info_from_type( *header ) );
+    DESERIALIZE( 1, metadata->salt_length = *header );
+    DESERIALIZE( 4, iterations = ( (unsigned long) header[0] << 24 |
+                                   (unsigned long) header[1] << 16 |
+                                   (unsigned long) header[2] << 8 |
+                                   (unsigned long) header[3] ) );
+    if( iterations <= INT_MAX )
+        metadata->iterations = iterations;
+    else
+        metadata->iterations = -1;
+    DESERIALIZE_CHECK( 8,
+                       payload_size = ( (unsigned long long) header[0] << 56 |
+                                        (unsigned long long) header[1] << 48 |
+                                        (unsigned long long) header[2] << 40 |
+                                        (unsigned long long) header[3] << 32 |
+                                        (unsigned long long) header[4] << 24 |
+                                        (unsigned long long) header[5] << 16 |
+                                        (unsigned long long) header[6] << 8 |
+                                        (unsigned long long) header[7] ),
+                       payload_size <= SIZE_MAX,
+                       "Payload too large" );
+    metadata->payload_size = payload_size;
```

Comment:
This line is causing failures on Windows due to the conversion from `unsigned long long` to `size_t`. Given the check above it should be fine, but the compiler is complaining about it.